### ISSUE TITLE
add variable name for including pre-made .config file

### ISSUE THREAD

**Issue by ghost** [freedomofpress/ansible-role-grsecurity#48]
I have a painstakingly curated .config with the grsec/pax options I want to use. As of now I can manually add a copy: to the install-grsec-kernel for it to be built automatically.

However I wonder if it be useful to seamlessly add it to the grsec playbook with a variable and automatic copy task, would this be possible?

Personally I find that unless someone really knows the precise kernel features that they know they need, leaving it to common users to figure out is an error prone process and that they would be better off with a strict kernel config file at first and only later open it up if needed for some feature they require.

This feature is probably only worthwhile for the install-grsec-kernel playbook. 

If this seems reasonable, something else I thought of considerations is the inclusion of arrays of possible GRKERNSEC profiles such as listed on https://en.wikibooks.org/wiki/Grsecurity/Appendix/Grsecurity_and_PaX_Configuration_Options#Usage_Type

some of which make a lot of sense for this playbook out of the box:

GRKERNSEC_CONFIG_SERVER=y
GRKERNSEC_CONFIG_VIRT_GUEST=y
GRKERNSEC_CONFIG_VIRT_VIRTUALBOX=y

which could be a single variable like:

{{ grkernsec_server_virt_guest_virtualbox}} (or something to that effect)

if users of this playbook intend to use it for servers or desktops ( {{ grkernsec_desktop}} ), they should be able to copy the playbook into a new directory and change the default profile definitions inside defaults/main.yml and be done with any required changes. Currently that is not possible because it requires manually updating the .config file on the system which is something we want users to avoid with orchestration.

**Comment by garrettr** [comment]
That sounds like it should be pretty easy to add as an optional feature; @conorsch what do you think?

**Comment by conorsch** [comment]
> However I wonder if it be useful to seamlessly add it to the grsec playbook with a variable and automatic copy task, would this be possible?

This is absolutely a good idea, and in fact would ease my own workflow. Right now what I'm doing is copying in the custom config file into the install role's `files` dir, naming in `config-foo`, then referencing it by setting `grsecurity_build_strategy: foo`. That works, but supporting a custom `src=` path would be more intuitive for ansible users. 

> which could be a single variable like:
> {{ grkernsec_server_virt_guest_virtualbox}} (or something to that effect)

Agreed that this would be a great feature, but I'm wary about the feasibility of a stable solution here. The build role runs `make olddefconfig` after copying the custom config file and before running the compile task, to ensure that newer kernel versions still work with aging kernel configs, by setting new or unreferenced kernel config options to the default values.

A few months ago I backed off of the fully automated approach you proposed @Andrew415, but I'm still interested in supporting it. I think in the meantime it would be beneficial to collect a variety of reviewed kernel configs and ship them as part of the role. A custom config file targeting general desktop usage would be wonderful. In a local branch I'm using a `digitaloceanminimal` config that rips out a lot of the drivers that ship with the vanilla kernel, which the `digitalocean` config leaves in place, to be intentionally generic.

**Comment by conorsch** [comment]
Managing multiple configs on a per-host basis can be quite irritating without a dedicated var, as @Andrew415 proposes. The safest approach for backwards compatibility is leveraging a `with_first_found` include for a custom config file, or falling back to one of the `grsecurity_build_strategy` vars that ship with the role. That way we can continue to support multiple common "vanilla" setups by shipping configs with the role, but also support arbitrary config files via a custom path.

**Comment by conorsch** [comment]
Reopening. The changes in #68 properly provide support for custom configs, but cause the `manual` logic to fail, since there's no lookup found for the "manual" config.

**Comment by conorsch** [comment]
False alarm, the changes in #68 seem sufficient. The manual build strategy completes without error on the master branch:

```
TASK [build-grsec-kernel : Advise on next steps for manually compiling kernel.] 
task path: /home/conor/freedomofpress/ansible-role-grsecurity/roles/build-grsec-kernel/tasks/prepare_source_directory.yml:35
ok: [grsec-build] => {
    "msg": "Build environment ready. Kernel source is available in /home/vagrant/linux/linux-4.7.6 . Run `make menuconfig` in that directory to configure the build, then `make-kpkg kernel_image` or similar to begin compiling the kernel.\n"
}

TASK [build-grsec-kernel : include] ********************************************
task path: /home/conor/freedomofpress/ansible-role-grsecurity/roles/build-grsec-kernel/tasks/main.yml:44
skipping: [grsec-build] => {"changed": false, "skip_reason": "Conditional check failed", "skipped": true}

TASK [build-grsec-kernel : include] ********************************************
task path: /home/conor/freedomofpress/ansible-role-grsecurity/roles/build-grsec-kernel/tasks/main.yml:49
skipping: [grsec-build] => {"changed": false, "skip_reason": "Conditional check failed", "skipped": true}

PLAY RECAP *********************************************************************
grsec-build                : ok=32   changed=5    unreachable=0    failed=0 
```

Tested with Ansible versions:
- 2.1.0
- 2.1.1
- 2.1.2

Closing.

**Comment by conorsch** [comment]
This issue has been flapping open and closed for a while now. Finally was able to reproduce the error on `master` branch. STR:

```
# set `grsecurity_build_strategy` to "digitalocean" in `extra_vars` block of Vagrantfile
vagrant up --provision grsec-build
```

The _manual_ build strategy still completes without error, but the custom file logic is currently broken. Patch incoming.
